doc/dev/backend-plugins.md

```diff
@@ -0,0 +1,271 @@
+# Proposal: Backend Plugins
+
+## Backend Contract
+
+There exists a _backend contract_ between `libelektra-kdb` and any plugin acting as a backend plugin.
+This contract sets, the order of the phases described above and defines the interaction between a backend plugin and `libelektra-kdb`.
+
+TODO: update get diagram; add cachecheck phase
+
+![Sequence of phases in `get` operation](kdbGet.svg)
+
+![Sequence of phases in `set` operation](kdbSet.svg)
+
+The diagrams above show the possible sequences of phases during a `get` and a `set` operation.
+For each of the phases of a `get` operation `libelektra-kdb` calls the backend plugin's `elektra<Plugin>Get` function once.
+Similarly, for the phases of a `set` operation `elektra<Plugin>Set` is called.
+
+The current phase is communicated to the backend plugin via the global keyset.
+The value of the key `system:/elektra/kdb/backend/phase` is always set to the current phase.
+
+### Operation `get`
+
+The `get` operation is mandatory and all backend plugins must implement it.
+
+#### Initialization Phase
+
+During the `init` phase the backend plugin is called with:
+
+- A key `parentKey` whose name is the root of the mountpoint configuration (e.g. `system:/elektra/mountpoints/system:\/hosts`) and whose value is an empty string.
+  The key name and value of this key are read-only.
+  The name of `parentKey` is chosen to make it easier for the plugin to produce good error messages.
+- A keyset `definition` containing the mountpoint definition.
+  To make things easier for the plugin, keys in `definition` are translated into cascading keys relative to `parentKey`.
```

Comment:
Is this needed? Why not simply cutting it?

Reply:
It makes the code in the backend plugin easier. For example we can do
```c
ksLookupByName (ksDefinition, "/path", 0);
```
instead of needing
```c
Key * lookup = keyDup (parentKey, KEY_CP_NAME);
keyAddName (lookup, "path");
ksLookup (ksDefinition, lookup, 0);
```

As a side effect is also saves memory.